accepted/2021/architecture-targeting.md

```diff
@@ -0,0 +1,496 @@
+
+The reason to keep portable apps as the default is that it is easier for non-container web developers that target a different operating system than the one they use locally for development. Portable apps are significantly worse for all other scenarios, particularly client.
+
+In short, portable apps should be an opt-in scenario. The user should make a decision on the extra app size and alternate folder structure that they will experience.
+
+Some people may argue that portable apps benefit new developers since they are simpler. That's potentially true, if they are deploying to a PaaS service that runs another operating system or uses a different CPU type. On the other hand, portable apps have the worst metrics (they can be significantly larger) such that their utility for the new developer experience is questionable. They are not our "best foot forward".
+
+## Proposal for behavior changes
+
+There are multiple proposed changes.
+
+### Enable RID-specific as default build type
+
+The default build type should be changed to the RID-specific framework-dependent model. This approach has the following benefits:
+
+- Apps are smaller (no unusable RID assets).
+- Apps likely build faster on slow drives (copy less assets to `bin` folder).
+- Apps startup faster (no need to do RID parsing/matching on startup).
+- All app assets are symmetric (the apphost matches other app assets and the runtime environment).
+- Simpler folder layout (there is none).
+
+The most significant beneficiary of this change would be containerized ASP.NET Core apps with RID-specific dependencies. We'd also get to delete a bunch of [architecture-specific Dockerfiles](https://github.com/dotnet/dotnet-docker/tree/main/samples/aspnetapp). A number of those Dockerfiles only exist due to the current architectural targeting CLI model.
+
+This change would resolve the request for [dotnet-publish with --use-current-rid or --infer-rid option](https://github.com/dotnet/sdk/issues/10449) without needing any additional syntax. This request aligns with the challenge we have with Dockerfiles.
+
+Portable framework-dependent apps are the default scenario today. As already stated, it's not a good default choice because it produces larger apps with niche benefits.
+
+### Enable gesture to build portable apps
+
+If rid-specific apps become the default, then we need a new gesture for portable apps. There are two (non-mutually-exclusive) models for that:
+
+- Use the existing `any` RID to mean portable.
+- Expose a new CLI arg and MSBuild property for portable apps, like  `--portable` and `<Portable>true</Portable>`.
+
+Recommendation: Start with using just the RID. Wait for feedback to enable more polished experiences to validate that portable is an important scenario.
```

Comment:
I believe we need the MSBuild option from day 1 or some alternative.

`dotnet build --runtime ... MyRepo.sln` fails today with `error NETSDK1134: Building a solution with a specific RuntimeIdentifier is not supported. If you would like to publish for a single RID, specifiy the RID at the individual project level instead.` and so you cannot actually decide to build an entire solution for a given RID. Doing this per project is slow and unmaintainable for larger projects.

Reply:
Interesting. I didn't know that. I didn't include `.sln` files in my experiments/research. Fail on that.

I take it that you'd be fine with `-r any` as long as it is respected on a `.sln` file. Yes?

Reply:
https://github.com/dotnet/designs/pull/233/commits/9a9528228faabd954e353b0efbd8269501194fe7

Reply:
> I take it that you'd be fine with -r any as long as it is respected on a .sln file. Yes?

Yes, provided you can build the entire solution as desired with a single command, I think that "fits the bill".

However, it's generally nice to be able to do just `dotnet build` or just `dotnet run` and have things work as expected and not have to pass command line args for it to be "correct" (assuming the project wants portable by default).